topomodelx/nn/combinatorial/hmc_layer.py

```diff
@@ -0,0 +1,1075 @@
+        elif self.update_func == "tanh":
+            return torch.nn.functional.tanh(message)
+
+    def attention(
+        self, message: torch.Tensor, A_p: torch.Tensor, a_p: torch.Tensor
+    ) -> torch.Tensor:
+        """Compute the attention matrix.
+
+        Parameters
+        ----------
+        message : torch.Tensor, shape=[n_messages, source_out_channels]
+            Message tensor. This is the result of the matrix multiplication
+            of the cochain matrix :math:`X` with the learnable weights
+            matrix :math:`W_p`.
+        A_p : torch.sparse, shape=[n_cells, n_cells]
+            Neighborhood matrix to the power p. Indicates how many paths of
+            lenght p exist from cell :math:`i` to cell :math:`j`.
+        a_p : torch.Tensor, shape=[2*source_out_channels, 1]
+            Learnable attention weight vector.
+
+        Returns
+        -------
+        att_p : torch.sparse, shape=[n_messages, n_messages].
+            Represents the attention matrix :math:`A_p`.
+        """
+        n_messages = message.shape[0]
+        source_index_i, source_index_j = A_p.coalesce().indices()
+        s_to_s = torch.cat([message[source_index_i], message[source_index_j]], dim=1)
+        e_p = torch.sparse_coo_tensor(
+            indices=torch.tensor([source_index_i.tolist(), source_index_j.tolist()]),
+            values=F.leaky_relu(
+                torch.matmul(s_to_s, a_p), negative_slope=self.negative_slope
+            ).squeeze(1),
+            size=(n_messages, n_messages),
+            device=self.get_device(),
+        )
+        att_p = (
+            torch.sparse.softmax(e_p, dim=1) if self.softmax else sparse_row_norm(e_p)
+        )
+        return att_p
+
+    def forward(
+        self, x_source: torch.Tensor, neighborhood: torch.Tensor
+    ) -> torch.Tensor:
+        r"""Compute forward pass.
+
+        The forward pass of the Higher Order Attention Block for squared
+        neighborhood matrices is defined as:
+
+        ..  math::
+            \text{HBS}_N(X) = \phi(\sum_{p=1}^{\text{m_hop}}(N^p \odot A_p) X
+            W_p ).
+
+        Parameters
+        ----------
+        x_source : torch.Tensor, shape=[n_cells, source_in_channels]
+            Cochain matrix representation :math:`X` whose rows correspond to
+            the signal features over each cell following the order of the cells
+            in :math:`\Sigma^s`.
+        neighborhood : torch.sparse, shape=[n_cells, n_cells]
+            Neighborhood matrix :math:`N`.
+
+        Returns
+        -------
+        _ : Tensor, shape=[n_cells, source_out_channels]
+            Output features of the layer.
+        """
+        message = [
+            torch.mm(x_source, w) for w in self.weight
+        ]  # [m-hop, n_source_cells, d_t_out]
+        # Create a torch.eye with the device of x_source
+        result = torch.eye(x_source.shape[0], device=self.get_device()).to_sparse_coo()
+
+        neighborhood = [
```

Comment:
Please don't overwrite a variable with another type, just use a different name. This is hard to read.

Reply:
@rballeba  this is confusing for me too. I tried to fix this but I am not sure what you are trying to do here in your implementation. Can you  please fix this PR from here? this is supposed to be easy for you since you coded this layer originally. Thanks!